### PR TITLE
Add ChainerX `FloorDivide`

### LINKED ISSUE #6344: Add ChainerX FloorDivide

### PR DESCRIPTION
**Pull request by asi1024**
Closes #6344.

### CODE REVIEW

File: chainerx_cc/chainerx/cuda/numeric.cuh [hunk 1]
```diff
@@ -50,5 +50,14 @@ __device__ inline T Sqrt(T x) {
 
 __device__ inline cuda::Float16 Sqrt(cuda::Float16 x) { return x.Sqrt(); }
 
+template <typename T>
+__device__ inline T FloorDivide(T x, T y) {
+    return x / y - (x % y < 0 ? T{1} : T{0});
```
[the remark this answers is not in the document]
Comment (asi1024) [reply chainerx_cc/chainerx/cuda/numeric.cuh]: `std::div` supports only `int32_t` and `int64_t` parameters.

File: chainerx_cc/chainerx/routines/math.cc [hunk 2]
```diff
@@ -258,8 +258,50 @@ Array Multiply(Scalar x1, const Array& x2) { return Multiply(x2, x1); }
...
+        NoBackpropModeScope scope{};
+        x1.device().FloorDivide(x1, x2, out);
+    }
+
+    {
+        BackwardBuilder bb{"floor_divide", {x1, x2}, out};
+        if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+            bt.Define([](BackwardContext& bctx) {
+                const Array& gout = *bctx.output_grad();
+                bctx.input_grad() = gout * Scalar{0, gout.dtype()};
```
Comment (asi1024) [inline chainerx_cc/chainerx/routines/math.cc]: Is there any better implementation?

File: chainerx_cc/chainerx/cuda/numeric.cuh (hunk 1 above)
[the remark this answers is not in the document]
Comment (niboshi) [reply chainerx_cc/chainerx/cuda/numeric.cuh]: How about specializing each type?
int8/16 ... cast is much lighter than division.
unsigned ... calculated in a single division.

File: chainerx_cc/chainerx/routines/math.cc (hunk 2 above)
Comment (niboshi) [reply chainerx_cc/chainerx/routines/math.cc]: Does this operation need backward implementation?

File: chainerx_cc/chainerx/cuda/numeric.cuh (hunk 1 above)
[the remark this answers is not in the document]
Comment (asi1024) [reply chainerx_cc/chainerx/cuda/numeric.cuh]: Fixed.

File: chainerx_cc/chainerx/routines/math.cc (hunk 2 above)
Comment (asi1024) [reply chainerx_cc/chainerx/routines/math.cc]: Removed.

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py [hunk 3]
```diff
@@ -209,6 +209,32 @@ def test_imul_scalar(xp, scalar, device, shape, dtype):
     return lhs
 
 
+# TODO(imanishi): Support and test bool dtype.
+@chainerx.testing.numpy_chainerx_array_equal(float16_rtol=1e-3)
+@pytest.mark.parametrize_device(['native:0', 'cuda:0'])
+def test_floordiv(xp, device, shape, numeric_dtype):
+    lhs = array_utils.create_dummy_ndarray(xp, shape, numeric_dtype)
+    rhs = xp.arange(1, lhs.size + 1, dtype=numeric_dtype).reshape(shape)
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: It seems negative numbers / fractional numbers are not tested?

File: chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu [hunk 4]
```diff
@@ -139,6 +140,57 @@ void CudaDevice::MultiplyAS(const Array& x1, Scalar x2, const Array& out) {
 
 namespace {
 
+__device__ int8_t FloorDivide(int8_t x, int8_t y) { return x / y - (x % y < 0 ? 1 : 0); }
```
Comment (niboshi) [inline chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu]: Doesn't CUDA have `std::div`? Why not write some comment?

**Comment by niboshi** [comment]
How about adding documentation?

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py [hunk 5]
```diff
@@ -209,6 +209,97 @@ def test_imul_scalar(xp, scalar, device, shape, dtype):
...
+    if (numeric_dtype in chainerx.testing.unsigned_dtypes and
+            (lhs_is_negative or rhs_is_negative)):
+        return chainerx.testing.ignore()
+    lhs = array_utils.create_dummy_ndarray(xp, shape, numeric_dtype)
+    rhs = xp.arange(1, lhs.size + 1, dtype=numeric_dtype).reshape(shape)
+    if lhs_is_negative:
+        lhs = -lhs
+    if rhs_is_negative:
+        rhs = -rhs
+    print(lhs, rhs)
```
Comment (hvy) [inline tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Debug print left in code.
```suggestion
```

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py [hunk 6]
```diff
@@ -209,6 +209,78 @@ def test_imul_scalar(xp, scalar, device, shape, dtype):
...
+    ([0, 1, 2, 3, 100, 101, 102, 103], -3),
+    ([-1, -2, -3, -4, -100, -101, -102, -103], -3),
+    ([0., 0.8, 1.6, 2.4, 100., 100.8, 101.6, 102.4], 1.2),
+    ([-0.8, -1.6, -2.4, -3.2, -100., -100.8, -101.6, -102.4], 1.2),
+    ([0., 0.8, 1.6, 2.4, 100., 100.8, 101.6, 102.4], -1.2),
+    ([-0.8, -1.6, -2.4, -3.2, -100., -100.8, -101.6, -102.4], -1.2),
+])
+@pytest.mark.parametrize_device(['native:0', 'cuda:0'])
+def test_floordiv(xp, lhs, rhs, device, numeric_dtype, is_module):
+    if numpy.array(lhs).dtype.kind != numpy.dtype(numeric_dtype).kind:
```
[the remark this answers is not in the document]
Comment (asi1024) [reply tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Fixed.

File: chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu [hunk 7]
```diff
@@ -139,6 +140,64 @@ void CudaDevice::MultiplyAS(const Array& x1, Scalar x2, const Array& out) {
...
 
+// CUDA does not have std::div.
+__device__ int8_t FloorDivide(int8_t x, int8_t y) { return x / y - ((y >= 0 ? x % y : -(x % y)) < 0 ? 1 : 0); }
+__device__ int16_t FloorDivide(int16_t x, int16_t y) { return x / y - ((y >= 0 ? x % y : -(x % y)) < 0 ? 1 : 0); }
+__device__ int32_t FloorDivide(int32_t x, int32_t y) { return x / y - ((y >= 0 ? x % y : -(x % y)) < 0 ? 1 : 0); }
+__device__ int64_t FloorDivide(int64_t x, int64_t y) { return x / y - ((y >= 0 ? x % y : -(x % y)) < 0 ? 1 : 0); }
+__device__ uint8_t FloorDivide(uint8_t x, uint8_t y) { return x / y; }
+__device__ float FloorDivide(float x, float y) {
+    float rem = std::fmod(x, y);
+    return (x - rem) / y - ((rem < 0 && y > 0) || (rem > 0 && y < 0) ? 1 : 0);
```
[the remark this answers is not in the document]
Comment (asi1024) [reply chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu]: The above suggestion changes the behavior if `x < 0 && rem == 0 && y > 0`.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins failure is irrelevant.

**Comment by niboshi** [comment]
LGTM

**Pull request merged**